Feature/badger db v3

evgmik opened the pull request:
Addresses #92

`badger v3.2011.1` made a few old options obsolete so they are removed.
I have adjusted the value log file to 10MB (from default 2GB), and memory table size to 10MB from default 128MB. If we do our job right we do not need that much anyway.

Another reason for this changes. `badger` allocate above values on hard drive, if user does not use modern FS which allow sparse files (with zeros). He or she would see 2GB of space disappeared before any operation with `brig` starts.
On modern FS, one would see *actual* allocation. Still 2GB file size looks scary to me.

So far, I see no negative effects from the badger version upgrade.

To to see compression improvements, the DB need to be garbage collected which can be done only at db.Close().

evgmik committed:
changing badger DB version v1.6.2 -> v3.2011.1

This upgrades the versions and removes options which were removed
upstream.

evgmik committed:
updating go.sum hashes

evgmik committed:
Merge branch 'develop' into feature/badger-db-v3

evgmik committed:
setting appropriate DB options and removing obsolete options

evgmik committed:
removed unused package

evgmik committed:
Merge branch 'fix/close-DBs-on-Quit' into feature/badger-db-v3

evgmik committed:
new option setting  WithNumVersionsToKeep(1) and reformatting

evgmik committed:
Merge branch 'develop' into feature/badger-db-v3, insisting on badger v3

evgmik committed:
added working badger GC, which closes and reopen DB

evgmik committed:
Merge branch 'develop' into feature/badger-db-v3

evgmik commented:
Looks like we ready to merge with develop.

I decided to skip GC for `gateway/db/db.go` it is rarely touched and is not poluted with obsolete records.

By the way, gateway db has a lot of interfaces which seems to be copy pasted from `catfs/db/database_badger.go`. Maybe there is a way to unite them, then they can use the same GC routine.

sahib commented:
> I decided to skip GC for gateway/db/db.go it is rarely touched and is not poluted with obsolete records.

Fine to me.

> By the way, gateway db has a lot of interfaces which seems to be copy pasted from catfs/db/database_badger.go. Maybe there is a way to unite them, then they can use the same GC routine.

Right. Might use badger also for as persistent cache for overlay caching. In that process I might add a generic KV-util, if I think the abstraction is worth the effort.

sahib commented on catfs/db/database_badger.go:
```diff
@@ -56,17 +55,95 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
+	var beforeGC uint64 = 0
+	var total uint64 = 0
+	var err error
+	log.Infof("Performin GC for badger DB in %s", opts.Dir)
```
typo: "Performing".

evgmik replied:
fixed, but there is a new patch which makes it obsolete.

sahib commented on catfs/db/database_badger.go:
```diff
@@ -56,17 +55,95 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
+				log.Debugf("badger DB in %s has nothing for GC",opts.Dir)
+			}
+			continue
+		}
```
What's the purpose of still running ``RunValueLogGC(0.5)`` after Close? Is it essential to call?

evgmik replied:
According to the Note in https://github.com/dgraph-io/badger/issues/767#issuecomment-485713746
the compactors,  which collect statistics for GC, are run only on DB `Close()` at least on small size DB.
So `Close` is needed for our DB size.

But your comment made me do more digging. So some things about `badger` DB GC collector was wrongly reversed engineered by me: `RunValueLogGC()` actually compacts on the fly, no need to `Close()` to shrink the DB. The problem that `RunValueLogGC` does not collect stats by itself, but uses prepared by compactors.
Which leads for necessity of one `Close()` to make sure that compactors are forced. Apparently, even `Close`
does not force it and need about 1 minute delay between DB modification and `Close` for compactors to kick in.

So new 3cc667c6 patch get rid of cycled Close/Open (it does it only once per `runGC` call), also I get rid
of the timer since the oldest DB flush and the modification flag.

If only there were reliable way to check if DB was modified, then we would not do unnecessary `Close()`. Though I noticed that `filesystem GC` constantly messing with the metadata DB even on idle `brig`.

sahib replied:
Thanks for the explanation. Code looks better to me now.

One simple to gather statistic we could display would be to display how long the GC run took. I'm a little worried that it blocks all of brigs operation for considerable time, even if there weren't many writes in that time period.

sahib commented on catfs/db/database_badger.go:
```diff
@@ -56,17 +55,95 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
+		}
+		// Badger people use os.FileInfo to get total size of lsm and vlog files
+		// This is not actual but allocated of hdd size
+		lsm, vlog := bdb.db.Size() // values could be stale even after Close/Open cycle, badger updates it once per minute
```
Do we really need all this complicated logic if it's only for outputting info messages?

sahib commented on catfs/db/database_badger.go:
```diff
@@ -323,6 +400,10 @@ func (db *BadgerDatabase) Flush() error {
 	db.txn = nil
 	db.haveWrites = false
+	if time.Now().After(db.gcNeededAt) && !db.needsGC {
+		db.gcNeededAt = time.Now().Add(120 * time.Second) // badger updates stats once per minute, we give ample time for stats to settle
```
Just a style thing: Please put the comments above the line they're referring to to avoid extra long lines.

evgmik replied:
done

evgmik committed:
spelling typo fixed

evgmik committed:
Simplified GC and get rid of Close/Open inside of a cycle

Reading badger DB issues list, yielded the following.
RunValueLogGC() does clean up online. But on small databases (150MB) is
not big enough, the only way to update stats for GC is to close DB.
see Note in https://github.com/dgraph-io/badger/issues/767#issuecomment-485713746

Based on this information the logic is redone, to call Close only if
RunValueLogGC did not succeed.

sahib commented on catfs/db/database_badger.go:
```diff
@@ -56,17 +51,60 @@ func NewBadgerDatabase(path string) (*BadgerDatabase, error) {
+	// to update stats after DB was modified. But eventually GC stats will be ready.
+	err := bdb.db.Close()
+	if err != nil {
+		log.Fatalf("Could not close DB in %s", opts.Dir)
```
I generally see calling ``log.Fatalf`` as bad practice, especially if it's not in quick & dirty programs and especially in library code (I have a story to tell here). I'm not fully sure what the proper error handling here is, but we could enter a retry loop when fail to close or open the database.

evgmik replied:
I tried to mitigate with 91d48265. Error in `Close` is not fatal, we can try later. If we cannot open, then we do ten attempts with timeouts and then Fail. We cannot run without DB. Also, fixed Close which did not check for an active transaction. GC could be done or running system, but Close definitely not.

I would love to hear this story :) I was always thinking that if something going wrong the program should faill with error message so user would send an angry bug report and we fix the problem. If we are running industrial process, then the story is different.

sahib replied:
> Error in Close is not fatal

Is it? What if it was partially closed and we're left with a zombie db conn?

> I would love to hear this story :)  If we are running industrial process, then the story is different.

Story is: We once used a [gpio library](https://github.com/brian-armstrong/gpio/pull/16/files) to check some pins on our device. Sadly I didn't look into the library implementation, so that fault is on me, but their error handling was "log & exit". Which is a bad strategy as a library. I noticed this first when the program didn't first run on my work machine (obvious cause was that the gpio was not available there), but at some point we made the grave mistake to update the Go version to 1.14 and all hell broke loose. All services reading this pin suddenly crashed randomly after the update. And why was that? Because since 1.14 ``select()`` might return ``EINTR`` (rightly so) when it's execution was interrupted by the scheduler, which might happen at any point. The gpio library was not prepared for this and just jumped off the cliff by calling ``os.Exit(1)``. Fix was easy, but what we can learn from this:

* Never ever ``os.Exit()`` in library code.
* Check the quality of the library you're using.
* Error handling is hard.

> I was always thinking that if something going wrong the program should faill with error message so user would send an angry bug report and we fix the problem.

That's fine for interactive programs. Long running programs (and brig daemon counts as one) should do this only as last resort since the error gets buried in some logs (in the best case). But not being able to open the main database is one such case, so https://github.com/sahib/brig/commit/91d482653845748445bcb0e91335de793bc09acc is fine with me.

evgmik replied:
I see now why it was a painful experience. Indeed `fail` is super special and should be used as a last line of defense.

sahib reviewed (commented):
There are two optional subjects that we should discuss before merging. Otherwise looks good to me :+1:

evgmik committed:
prevent Closing DB if we still trasaction open, trying to reopen DB if we failed a few times

evgmik committed:
Added GC elapse time report

evgmik commented:
I have added GC elapsed time the worst time is about 170mS after big cleanup. Typical time is about 15mS with Close/Open of DB before patch f7ae738.

With patch f7ae738 GC is very quick, since it does not do unnecessary (Close/Open) work on GC if stats are known to be not refreshed

evgmik committed:
added tracker of DB updates time.

Run GC only after some DB update with a delay to give time for GC
relevant stats to be updated by `badger`

This prevent doing useless DB Close/Open for the cases when we know for sure
that no changes in DB were done after last garbage collection.

evgmik committed:
gofmt

evgmik committed:
Retring DB closing several times before reporting failure

extra change, retry loop done with
```
for retries := 0; retries < 10; retries++ {
```

evgmik commented:
You are right:  `for retries := 0; retries < 10; retries++ { ... }` is much cleaner. Fixed in 5e5821c1.
In the same patch retrying `db.Close`. You are right yet again, it could be left in the half broken state after partial `Close`
